spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -779,4 +761,56 @@ private MessageHandlerMethodFactory createDefaultMessageHandlerMethodFactory() {
 
 	}
 
+	private class TypeMetadataHolder {
```

Comment:
Looks like we don't need this intermediate, singleton-aware holder.
We really just can `compute` `TypeMetadata` if it absent in the `typeCache` map.
I'd even wouldn't worry a lot about concurrent override since all the info from the `Class` is idempotent.